### PR TITLE
protocol parser: rdp

### PR DESCRIPTION
**Pull request by lmzke**
Continuation of #4088.

Initial implementation of issue [2314](https://redmine.openinfosecfoundation.org/issues/2314):
1. Add protocol parser for RDP
2. Add transactions for RDP negotiation
3. Add eve logging of transactions
4. Bumped rustc versions in Travis

Travis CI rustc minimum bumped to v1.32.0 (2019-01-17), the minimum version needed by dependencies.  Latest known-good version bumped to v1.36.0 (2019-07-04), which is also the current rustc stable.

@jasonish mentioned in #4088 that the master/5.x branches are not able to leave minimum v1.24.1 (2018-03-01) yet.  This PR may need to block until then.

New contributor, all feedback welcome.

Sample output, based on [wireshark RDP-002.pcap](https://wiki.wireshark.org/SampleCaptures?action=AttachFile&do=view&target=RDP-002.pcap.gz):

```json
{
  "timestamp": "2007-10-24T18:58:09.119601-0400",
  "flow_id": 2031235224416398,
  "pcap_cnt": 5,
  "event_type": "rdp",
  "src_ip": "10.226.24.52",
  "src_port": 3389,
  "dest_ip": "10.226.41.226",
  "dest_port": 10446,
  "proto": "TCP",
  "rdp": {
    "tx_id": 0,
    "event_type": "initial_request",
    "cookie": "A70067"
  }
}
{
  "timestamp": "2007-10-24T18:58:09.119844-0400",
  "flow_id": 2031235224416398,
  "pcap_cnt": 6,
  "event_type": "rdp",
  "src_ip": "10.226.41.226",
  "src_port": 10446,
  "dest_ip": "10.226.24.52",
  "dest_port": 3389,
  "proto": "TCP",
  "rdp": {
    "tx_id": 1,
    "event_type": "initial_response"
  }
}
{
  "timestamp": "2007-10-24T18:58:09.120861-0400",
  "flow_id": 2031235224416398,
  "pcap_cnt": 7,
  "event_type": "rdp",
  "src_ip": "10.226.24.52",
  "src_port": 3389,
  "dest_ip": "10.226.41.226",
  "dest_port": 10446,
  "proto": "TCP",
  "rdp": {
    "tx_id": 2,
    "event_type": "connect_request",
    "client": {
      "version": "v5",
      "desktop_width": 1152,
      "desktop_height": 864,
      "color_depth": 15,
      "keyboard_layout": "en-US",
      "build": "Windows XP",
      "client_name": "ISD2-KM84178",
      "keyboard_type": "enhanced",
      "function_keys": 12,
      "product_id": 1,
      "capabilities": [
        "support_errinfo_pdf"
      ],
      "id": "55274-OEM-0011903-00107"
    },
    "channels": [
      "rdpdr",
      "cliprdr",
      "rdpsnd"
    ]
  }
}
{
  "timestamp": "2007-10-24T18:58:09.121164-0400",
  "flow_id": 2031235224416398,
  "pcap_cnt": 8,
  "event_type": "rdp",
  "src_ip": "10.226.41.226",
  "src_port": 10446,
  "dest_ip": "10.226.24.52",
  "dest_port": 3389,
  "proto": "TCP",
  "rdp": {
    "tx_id": 3,
    "event_type": "connect_response"
  }
}
```

### CODE REVIEW

**Comment by jasonish** [comment]
> New contributor, all feedback welcome.

Thanks for the contribution.  Due to summer holidays it may be a week or more before we can give this a good look.

**Comment by pevma** [comment]
info note: I am putting it for some side/live testing. Will feedback if something jumps out right away.

**Comment by norg** [comment]
One thing you need to make sure is that you also include documentation for the parser/keywords which is one of the requirements we add with the PR template. Another thing is the necessary contributor agreement.

So while waiting for feedback you might want to add documentation and use the github PR template from us for the next PR :)

**Comment by lmzke** [comment]
Template used in #4088 (but incorrectly relied on n/a).  Will make sure to update `doc/userguide/output/eve/eve-json-format.rst`.

**Comment by victorjulien** [comment]
The new dependencies are all licensed MIT (plus either APL2 or "Unlicensed"), so compatible to GPL2.

File: .travis.yml [hunk 1]
```diff
@@ -118,42 +118,42 @@ matrix:
         - DO_CHECK_SETUP_SCRIPTS="yes"
         - DO_DISTCHECK="yes"
     # Linux, gcc, Rust (auto detect).
-    # - Rust 1.31.0, the latest known working version.
```
Comment (victorjulien) [inline .travis.yml]: this travis update should be a separate commit

**Comment by victorjulien** [comment]
@chifflier are you interested in doing a review of this? Full would be great, but only the parts that interact with your tls-parser and x509-parser crates would be helpful too.

File: src/output-json-rdp-rust.c [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+#include "util-debug.h"
+#include "util-byte.h"
+#include "output.h"
+#include "output-json.h"
+#include "app-layer.h"
+#include "app-layer-parser.h"
+#include "app-layer-rdp-rust.h"
+#include "output-json-rdp-rust.h"
+
+#if defined(HAVE_LIBJANSSON) && defined(HAVE_RUST)
```
Comment (jlucovsky) [inline src/output-json-rdp-rust.c]: Both of these are mandatory so the guards are no longer required.

**Comment by chifflier** [comment]
> @chifflier are you interested in doing a review of this? Full would be great, but only the parts that interact with your tls-parser and x509-parser crates would be helpful too.

Sure, I'll have a look. RDP is quite a complex protocol, so it is an interesting PR proposal.
I hadn't noticed first that {tls,x509}-parsers were used :)

File: rust/src/rdp/parser.rs [hunk 3]
```diff
@@ -0,0 +1,1559 @@
...
+    pub dst_ref: u16,
+    pub src_ref: u16,
+    pub class: u8,
+    pub options: u8,
+    pub negotiation_from_server: Option<NegotiationFromServer>,
+}
+
+/// variants of a server negotiation
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
```
Comment (chifflier) [inline rust/src/rdp/parser.rs]: Same here: why `repr(C)`?

File: rust/src/rdp/parser.rs [hunk 4]
```diff
@@ -0,0 +1,1559 @@
...
+
+/// x224-spec, section 13.7
+#[derive(Clone, Debug, PartialEq)]
+pub struct X223Data {
+    pub child: X223DataChild,
+}
+
+/// variants that an x.223 data message can hold
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
```
Comment (chifflier) [inline rust/src/rdp/parser.rs]: same

File: rust/src/rdp/parser.rs [hunk 5]
```diff
@@ -0,0 +1,1559 @@
...
+
+/// rdp-spec, section 2.2.1.3.2
+#[derive(Clone, Debug, PartialEq)]
+pub struct McsConnectRequest {
+    pub children: Vec<McsConnectRequestChild>,
+}
+
+/// variants that an mcs connection message can hold
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
```
Comment (chifflier) [inline rust/src/rdp/parser.rs]: same

File: rust/src/rdp/parser.rs [hunk 6]
```diff
@@ -0,0 +1,1559 @@
...
+/// rdp-spec, section 2.2.1.4
+#[derive(Clone, Debug, PartialEq)]
+pub struct McsConnectResponse {}
+
+// ==================
+
+/// parser for t.123 and children
+/// t.123-spec, section 8
+pub fn parse_t123_tpkt(input: &[u8]) -> IResult<&[u8], T123Tpkt> {
+    let (i1, _version) =
```
Comment (chifflier) [inline rust/src/rdp/parser.rs]: `try_parse` can be replaced by the `?` operator:
```rust
let (i1, _version) = verify!(input, be_u8, |x| x == TpktVersion::T123 as u8)?;
```

File: rust/src/rdp/parser.rs [hunk 7]
```diff
@@ -0,0 +1,1559 @@
...
+        Err(nom::Err::Error(error_position!(
+            input,
+            ErrorKind::Custom(RDP_NOT_X224_CLASS_0_ERROR)
+        )))
+    }
+}
+
+// rdp-spec, section 2.2.1.1.1
+named!(pub parse_rdp_cookie<RdpCookie>,
+    do_parse!(
```
Comment (chifflier) [inline rust/src/rdp/parser.rs]: Info: rustfmt tends to mess up indentation in `do_parse` macros. However, if you use '{' instead of '(', it will not, and the macro can be nicely indented:
```rust
do_parse! {
  _key: verify!(...) >>
  _name: verify!(...) >>
```

Also (minor), `"Cookie: ".as_bytes()` can be written `b"Cookie: "`

File: rust/src/rdp/parser.rs [hunk 8]
```diff
@@ -0,0 +1,1559 @@
...
+    let child: T123TpktChild = match opt3 {
+        Some(x) => x,
+        None => T123TpktChild::Raw(data.to_vec()),
+    };
+
+    return Ok((i4, T123Tpkt { child }));
+}
+
+/// rdp-spec, section 2.2.1.1
+pub fn parse_x224_connection_request(
```
Comment (chifflier) [inline rust/src/rdp/parser.rs]: Parser functions should not be `pub` if not exported

File: rust/src/rdp/parser.rs [hunk 9]
```diff
@@ -0,0 +1,1559 @@
...
+
+/// t.123-spec, section 8
+#[derive(Clone, Debug, PartialEq)]
+pub struct T123Tpkt {
+    pub child: T123TpktChild,
+}
+
+/// variants that a t.123 tpkt can hold
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
```
[the remark this answers is not in the document]
Comment (lmzke) [reply rust/src/rdp/parser.rs]: Mistakenly believed it was useful to the engine overall.  Will remove (multiple locations).

File: rust/src/rdp/parser.rs (hunk 7 above)
Comment (lmzke) [reply rust/src/rdp/parser.rs]: Good tip for `do_parse` and rustfmt!  Will incorporate both suggestions.

**Comment by lmzke** [comment]
> @chifflier reviewed 8 hours ago

Beyond the code of this PR, my Rust experience is a bit limited.  I really appreciate all (on-going) suggestions: major, minor, and style.  Plus it's definitely not every day that you get the author of Rusticata and co-author of the parser whitepaper looking at your code!

File: rust/src/rdp/rdp.rs [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+
+const PARSER_NAME: &'static [u8] = b"rdp\0";
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_rdp_register_parser() {
+    let default_port = std::ffi::CString::new("[3389]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_TCP,
```
[the remark this answers is not in the document]
Comment (lmzke) [reply rust/src/rdp/rdp.rs]: > There seems to be multiple versions of RDP (4.0 to 10.0 ...

Early in the session, the client announces its version and the parser captures and logs that.  For our purposes (negotiation tracking), all versions are supported.  In actuality, those new/different capabilities occur later in the session, beyond our visibility.

See also section 2.2.1.3.2, "version", of https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/00f1da4a-ee9c-421a-852f-c19f92343d73.

In the sample output, e.g., it is (`.[2].rdp.client.version`) `"v5"`.  (Possible values include `"v4"`, `"v10.6"`, and `"v10.7"`, among others.)  Apologies that I have not documented the JSON schema for your easy review, I have that action.

[the remark this answers is not in the document]
Comment (lmzke) [reply rust/src/rdp/rdp.rs]: > IIRC, RDP can be both TCP and UDP - is supporting TCP only intentional

RDPUDP/RDPEUDP was off my radar.  https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeudp/2744a3ee-04fb-407b-a9e3-b3b2ded422b1  It uses its own TCP/IP connection management layer, and is not (simply) TPKT like RDP.  Let me take a look and reply again.

[the remark this answers is not in the document]
Comment (lmzke) [reply rust/src/rdp/rdp.rs]: Following up:

My current understanding is that RDP-UDP will sometimes be utilized after the RDP connection is fully established.  e.g., "After the connection has been secured and the Licensing phase has run to completion, the server can choose to initiate multi-transport connections. The Initiate Multitransport Request PDU is sent by the server to the client and results in the out-of-band creation of a multitransport connection using messages from the RDP-UDP, TLS, DTLS, and multitransport protocols..." (MS-RDPBCGR, section 1.3.1.1).  In this situation, the preliminary TCP-based RDP negotiation of this PR should correctly parse and log.

There are two transports for RDP over UDP: 1) RDP-UDP Forward Error Correction (FEC) reliable transport, and 2) RDP-UDP FEC lossy transport.  (MS-RDPBCGR, section 2.2.1.3.8)  FEC is present in both transports.

For any non-multi-transport case - i.e., if RDP-UDP is used for an actual RDP session negotiation - the FEC transport itself would have to be parsed, tracked, and reassembled first.  The resulting payload would not be TPKT, which is what results from TCP reassembly.  The TCP-based RDP parser of this PR could therefore not be used as a drop-in replacement.  Some types and functions in the RDP parser could be reused, as applicable.

My request/recommendation is that this PR (and associated code contribution) continue to focus only on RDP over TCP.

[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/rdp/rdp.rs]: Agreed. This functionality can be merged w/o UDP support. Think it would be good to open a feature ticket for that in our redmine.

**Comment by victorjulien** [comment]
Please submit a new PR with the requested updated, thanks!
